rgw/admin/errors.go

```diff
@@ -0,0 +1,268 @@
+// report that an index repair failed
+func IsIndexRepairFailed(err error) bool {
+	return reasonForError(err) == errIndexRepairFailed
+}
+
+// IsBucketNotEmpty returns a boolean indicating whether the error is known to
+// report that a bucket is not empty
+func IsBucketNotEmpty(err error) bool {
+	return reasonForError(err) == errBucketNotEmpty
+}
+
+// IsObjectRemovalFailed returns a boolean indicating whether the error is known to
+// report that the removal of an object failed
+func IsObjectRemovalFailed(err error) bool {
+	return reasonForError(err) == errObjectRemovalFailed
+}
+
+// IsBucketUnlinkFailed returns a boolean indicating whether the error is known to
+// report that the server was unable to unlink bucket to specified user
+func IsBucketUnlinkFailed(err error) bool {
+	return reasonForError(err) == errBucketUnlinkFailed
+}
+
+// IsBucketLinkFailed returns a boolean indicating whether the error is known to
+// report that the server was unable to link bucket to specified user
+func IsBucketLinkFailed(err error) bool {
+	return reasonForError(err) == errBucketLinkFailed
+}
+
+// IsNoSuchObject returns a boolean indicating whether the error is known to
+// report that the specified object does not exist
+func IsNoSuchObject(err error) bool {
+	return reasonForError(err) == errNoSuchObject
+}
+
+// IsIncompleteBody returns a boolean indicating whether the error is known to
+// report that either bucket was not specified for a bucket policy request or
+// bucket and object were not specified for an object policy request.
+func IsIncompleteBody(err error) bool {
+	return reasonForError(err) == errIncompleteBody
+}
+
+// IsNoSuchCap returns a boolean indicating whether the error is known to
+// report that the given  user does not possess specified capability
+func IsNoSuchCap(err error) bool {
+	return reasonForError(err) == errNoSuchCap
+}
+
+// IsInternalerror returns a boolean indicating whether the error is known to
+// report that the server suffered from an internal error
+func IsInternalerror(err error) bool {
+	return reasonForError(err) == errInternalerror
+}
+
+// IsAccessDenied returns a boolean indicating whether the error is known to
+// report that access to server is denied
+func IsAccessDenied(err error) bool {
+	return reasonForError(err) == errAccessDenied
+}
+
+// IsNoSuchBucket returns a boolean indicating whether the error is known to
+// report that the given bucket does not exist
+func IsNoSuchBucket(err error) bool {
+	return reasonForError(err) == errNoSuchBucket
+}
+
+// IsNoSuchKey returns a boolean indicating whether the error is known to
+// report that the given access key does not exist
+func IsNoSuchKey(err error) bool {
+	return reasonForError(err) == errNoSuchKey
+}
+
+// IsInvalidArgument returns a boolean indicating whether the error is known to
+// report that an invalid argument was passed to the server
+func IsInvalidArgument(err error) bool {
+	return reasonForError(err) == errInvalidArgument
+}
```

Comment:
I don't like using sentinel errors because it means that errors become part of the public API. Even though the code looks simpler it creates a tight relationship between the API and the consumer of the API. Using opaque error are more appealing as they remove this dependency and the caller only need to know in real human word the error he/she is looking for.

Reply:
Honestly, I feel like having `errors.Is(err, admin.ErrNoSuchKey)` or ` admin.IsNoSuchKey(err)` are the same thing. I don't see how the functions would create even more coupling. From my perspective, having the function is more user-friendly than having a sentinel error type. It's all about usability.

Reply:
Yeah, the "even more coupling" was meant in a way like not just a variable name, but a whole function signature, but that was a secondary point anyway. The main point was: it's certainly not less coupling than sentinels. And what is perceived as "user-friendly" is really depending on the context. Given that all other packages in go-ceph are using sentinel errors, it would for people like _me_ at least certainly be less intuitive and user-friendly, if I have to look up special functions to do error checking, instead of using the standard pattern provided by the go standard libraries. User-friendliness is also about consistency.

Reply:
It's the first time you are mentioning the consistency argument. I'm all for consistency, if the project already uses sentinel errors I'll do the same for this package.